ui/js/controllers/tcjobactions.js

```diff
@@ -0,0 +1,91 @@
+"use strict";
+
+treeherder.controller('TCJobActionsCtrl', [
+    '$scope', '$http', '$uibModalInstance', 'ThResultSetStore',
+    'ThJobDetailModel', 'thTaskcluster', 'ThTaskclusterErrors',
+    'thNotify', 'job', 'repoName', 'resultsetId',
+    function($scope, $http, $uibModalInstance, ThResultSetStore,
+             ThJobDetailModel, thTaskcluster, ThTaskclusterErrors, thNotify,
+             job, repoName, resultsetId) {
+        let originalTask;
+        $scope.input = {};
+
+        $scope.cancel = function() {
+            $uibModalInstance.dismiss('cancel');
+        };
+
+        $scope.updateSelectedAction = function() {
+            $scope.input.jsonPayload = JSON.stringify(jsonSchemaDefaults($scope.input.selectedAction.schema), null, 4);
+        };
+
+        $scope.triggerAction = function() {
+            $scope.triggering = true;
+
+            let tc = thTaskcluster.client();
+            let task = _.clone($scope.input.selectedAction.task);
+
+            ['created', 'deadline', 'expires'].map(function(dateProp) {
+                task[dateProp] = tc.fromNow(task[dateProp]['$fromNow']);
+            });
+            task.payload.env['ACTION_INPUT'] = $scope.input.jsonPayload;
+            task.payload.env['ACTION_PARAMETERS'] = null; // fixme: unclear what to set this to
+            task.payload.env['ACTION_TASK'] = null; // fixme: what is this supposed to do?
+            task.payload.env['ACTION_TASK_ID'] = JSON.stringify(job.taskcluster_metadata.task_id);
+            task.payload.env['ACTION_TASK_GROUP_ID'] = JSON.stringify(originalTask.taskGroupId);
```

Comment:
Basically, you cannot may any assumptions on the contents of `actions.json` other than those specified in the specification. In this case you some very specific assumptions about the contents and structure if the task template.
We want to decouple not link things tightly...

Anyways, the render stuff should be pretty solid... I'd be happy to help fix any bugs in both template and the `render` function.